### ISSUE TITLE
Excessive memory consumption

### ISSUE THREAD

**Issue by monodhs** [NLnetLabs/nsd#202]
Hello all,

Given the following nsd.conf for an NSD of version 4.1.26

```
server:
zonesdir: "/some/path/NSD"
zonelistfile: "/some/path/NSD/zone.list"
pidfile: "/some/path/NSD/PID"
xfrdfile: ""
do-ip4: yes
do-ip6: no
ip-address: 10.1.2.3
ip-address: 127.0.0.1
port: 53
ip-freebind: yes
ip-transparent: no
chroot: ""
database: ""
server-count: 1
reuseport: yes
rrl-ratelimit: 0
rrl-whitelist-ratelimit: 0
rrl-size: 1
logfile:        ""
log-time-ascii: no
verbosity: 2
username: ""
nsid: deadbeef
identity: deadbeef
version: 1
hide-version: yes
zonefiles-check: no
minimal-responses: no
refuse-any: yes
round-robin: no
include: "/some/path/NSD/zone:*"
remote-control:
control-enable: yes
control-interface: "/some/path/NSD-C"
```
with
```
zone:
name: "example.com."
zonefile: "/some/path/example.com."
provide-xfr: 1.2.3.4 NOKEY
provide-xfr: 2.3.4.5 NOKEY
provide-xfr: 3.4.5.6 NOKEY
notify: 1.2.3.4 NOKEY
notify: 2.3.4.5 NOKEY
notify: 3.4.5.6 NOKEY
notify-retry: 5
```
and
```
zone:
name: "another.example.com."
zonefile: "/some/path/another.example.com."
provide-xfr: 1.2.3.4 NOKEY
provide-xfr: 2.3.4.5 NOKEY
provide-xfr: 3.4.5.6 NOKEY
notify: 1.2.3.4 NOKEY
notify: 2.3.4.5 NOKEY
notify: 3.4.5.6 NOKEY
notify-retry: 5
```
for ``zone:example.com.`` and ``zone:another.example.com.``, respectively, I have seen NSD consuming more than 83 MB of memory right after startup, which has even increased to close to 98 MB after 10 hours of runtime by now.

Both zone files have not changed since startup, are less than 47 kB and 23 kB in size with 226 records (including all DNSSEC types), were signed using ``ldns-signzone``, and are verified and complete according to ``ldns-verify-zone``. Also, no zones were added to the server or deleted from it.

Traffic to the server is negligible, as it is operated as a Hidden Master, only being asked by its secondaries every four hours for SOA refresh.

Is such a high memory footprint as reported here to be expected for normal operation with NSD?

I had already managed to reduce the memory footprint from a meaty 129 MB right after startup by setting ``rrl-size: 1`` instead of leaving that set at its default, while still having RRL disabled (or so I assume), but after only 10 hours in and with almost 100 MB of memory claimed, I am more than certain that there must be other ways to curb NSD's hunger for core.

For reference, the same zone contents and traffic pattern operated with PowerDNS requires less than a thenth of the resources consumed by NSD with the above configuration.

Thanks!

**Comment by bjovereinder** [comment]
Hi,

Thank you for question.

We recently added options to reduce memory usage.  For its operation NSD allocates some buffers, but in some cases this is not necessary, e.g. the way you want to run NSD as a hidden primary name server.

The feature has not yet been released with the latest NSD version 4.3.9, but will likely be part of the next NSD release.

Best,

-- Benno

**Comment by monodhs** [comment]
I take this as a _yes_, then, that such memory hunger is by design.

No further questions, your Honour.
